### DISCUSSION TITLE
Problem in Modelling if-else-end statements

### DISCUSSION THREAD

**Discussion by HBA-DTT** [yalmip/YALMIP#1413] [Optimization modelling]
Dear prof.,
I am encountering difficulties while modeling with if-else-end statements. Here are the mathematical formulas I am working with:
<img width="411" alt="图片1" src="https://github.com/yalmip/YALMIP/assets/78578435/8a7b402f-8338-4ed7-bfd9-a2fc4439e575">
Here is my code:
~~~
clc
clear

% 0 <= x <= 11500 , y = 405.72*x
% 115000 <= x <= 200000 , y = 405.72*11500 + 554.9*(x - 115000)
% x >= 200000 , y = 405.72*11500 + 554.9*(200000 - 115000) + 604.69*(x - 200000)

%% data
Qit_in = [124 114 172 182 269 328 403 408 453 460 428 382 283 257 249 269 271 365 450 500 456 422 272 133];
Qit_max = 1.15*Qit_in;
Qit_min = 0.85*Qit_in;      


%% variables
y = sdpvar(1);
Qit = sdpvar(1,24);
d = binvar(1,3);
x = 30*sum(Qit);

%% Constraints
Constraints = [];
Constraints = [Constraints, Qit >= Qit_min , Qit <= Qit_max];
Constraints = [Constraints, sum(d) == 1];
Constraints = [Constraints, implies(d(1) , [0 <= x <= 115000 , y == 405.72*x] )];
Constraints = [Constraints, implies(d(2) , [115000 <= x <= 200000 , y == 405.72*11500 + 554.9*(x - 115000)] )];
Constraints = [Constraints, implies(d(3) , [x >= 200000 , y == 405.72*11500 + 554.9*(200000 - 115000) + 604.69*(x - 200000)] )];
Constraints = [Constraints, x >= 0 , x <= 2e6 , y >= 0 , y <= 1e10];

% gurobi
ops = sdpsettings('verbose', 3, 'solver', 'cplex','showprogress',1);
sol = optimize(Constraints, [], ops);
if sol.problem == 0
    disp('success');
else
    disp('something wrong');
    yalmiperror(sol.problem)
end
~~~
However, the results indicate that the model is infeasible. For debugging purposes, I have considered adding slack variables：
~~~
clc
clear

% 0 <= x <= 11500 , y = 405.72*x
% 115000 <= x <= 200000 , y = 405.72*11500 + 554.9*(x - 115000)
% x >= 200000 , y = 405.72*11500 + 554.9*(200000 - 115000) + 604.69*(x - 200000)

%% data
Qit_in = [124 114 172 182 269 328 403 408 453 460 428 382 283 257 249 269 271 365 450 500 456 422 272 133];
Qit_max = 1.15*Qit_in;
Qit_min = 0.85*Qit_in;      


%% variables
y = sdpvar(1);
Qit = sdpvar(1,24);
d = binvar(1,3);
x = 30*sum(Qit);
slack = sdpvar(1,3);

%% Constraints
Constraints = [];
Constraints = [Constraints, Qit >= Qit_min , Qit <= Qit_max];
Constraints = [Constraints, sum(d) == 1];
Constraints = [Constraints, implies(d(1) , [0 <= x <= 115000 , y == 405.72*x + slack(1)] )];
Constraints = [Constraints, implies(d(2) , [115000 <= x <= 200000 , y == 405.72*11500 + 554.9*(x - 115000) + slack(2)] )];
Constraints = [Constraints, implies(d(3) , [x >= 200000 , y == 405.72*11500 + 554.9*(200000 - 115000) + 604.69*(x - 200000) + slack(3)] )];
Constraints = [Constraints, x >= 0 , x <= 2e6 , y >= 0 , y <= 1e10];

% gurobi
ops = sdpsettings('verbose', 3, 'solver', 'cplex','showprogress',1);
sol = optimize(Constraints, sum(abs(slack)), ops);
if sol.problem == 0
    disp('success');
else
    disp('something wrong');
    yalmiperror(sol.problem)
end
~~~
The model is feasible now, but I don't know why there must be slack variables to make the model feasible.
I would appreciate any assistance or guidance you can provide in resolving this issue. Thank you for your time and attention.

**Comment by johanlofberg** [comment]
Runs without issues here. If you really have infeasibility issues I would suspect cplex to have a bug since the problem is trivial and even solved by YALMIPs internal bnb solver in the root node (find a presolve option in cplex and turn it off, very common reason for bugs. or even better, dump cplex and use gurobi/xpress/mosek instead)

`
 sol = optimize(Constraints, [], ops)
sol = 

  struct with fields:

    yalmipversion: '20230622'
    matlabversion: '9.6.0.1072779 (R2019a)'
       yalmiptime: 0.1339
       solvertime: 0.0291
             info: 'Successfully solved (Successfully solved (INTLINPROG))'
          problem: 0

>> x
Linear scalar (real, 24 variables)
Current value: 195075
Coefficients range: 30 to 30
>> y
Linear scalar (real, 1 variable)
Current value: 49099397.5
Coefficients range: 1 to 1
>> d
Linear matrix variable 1x3 (full, real, binary, 3 variables)
Values in range [0,1], 1 nonzeros
Coefficient range: 1 to 1
`
